### PR TITLE
allow custom initialization of the HandlerType of the LambdaRuntime

### PR DESCRIPTION
**Pull request by tomerd**
Motivation:  Provide the flexibility for custom initialization of the HandlerType as this will often be required by higher level frameworks.

Modifications:
* Modify the LambdaRuntime type to accept a closure to provide the handler rather than requiring that it is provided by a static method on the Handler type
* Update downstream code to use HandlerProvider
* Update upstream code to support passing Hanlder Type of Handler Provider
* Add and update tests

Originally suggested and coded by @tachyonics in https://github.com/swift-server/swift-aws-lambda-runtime/pull/308

### CODE REVIEW

**Comment by tomerd** [comment]
@tachyonics wdyt?

**Comment by tachyonics** [comment]
> @tachyonics wdyt?

This will work for our use case although not quite as convenient as what I had. My only concern is that `handlerProvider` returns a `ByteBufferLambdaHandler` which has a `makeHandler` function requirement that will not actually get used. In our code we can just not implement that method but conceptually it seems a bit odd at the runtime level.

**Comment by tomerd** [comment]
@tachyonics wdyt about https://github.com/swift-server/swift-aws-lambda-runtime/pull/310/commits/792d1c1adb0f49704a28f9093e87b618d48b5f5b? I don love the name `CoreByteBufferLambdaHandler` but hopefully this address the need better?

**Review by tachyonics — APPROVED** [review]
@tomerd I am not really a fan of the name either but otherwise the changes look good and are more consistent.

**Comment by tomerd** [comment]
thanks @tachyonics, so now to the hard part - choosing a name.

cc @fabianfett @sebsto @yim-lee @ktoso

any good ideas?

**Comment by tachyonics** [comment]
> so now to the hard part - choosing a name

@tomerd `NonInitializingByteBufferLambdaHandler`?

**Comment by tomerd** [comment]
@tachyonics updated to `NonInitializingByteBufferLambdaHandler`. wdyt?

File: Sources/AWSLambdaRuntimeCore/LambdaRuntime.swift [hunk 1]
```diff
@@ -226,8 +267,79 @@ public enum LambdaRuntimeFactory {
...
+            handlerProvider: Handler.makeHandler(context:),
+            eventLoop: eventLoop,
+            logger: logger
+        )
+    }
+
+    /// Create a new `LambdaRuntime`.
+    ///
+    /// - parameters:
+    ///     - handlerProvider: A provider of the ``CoreByteBufferLambdaHandler``  the `LambdaRuntime` will manage.
```
Comment (tachyonics) [inline Sources/AWSLambdaRuntimeCore/LambdaRuntime.swift:304]: Should this be `NonInitializingByteBufferLambdaHandler`?

File: Sources/AWSLambdaRuntimeCore/Lambda.swift [hunk 2]
```diff
@@ -85,6 +85,22 @@ public enum Lambda {
     internal static func run(
         configuration: LambdaConfiguration = .init(),
         handlerType: (some ByteBufferLambdaHandler).Type
+    ) -> Result<Int, Error> {
+        Self.run(configuration: configuration, handlerProvider: handlerType.makeHandler(context:))
+    }
+
+    /// Run a Lambda defined by implementing the ``ByteBufferLambdaHandler`` protocol.
```
Comment (tachyonics) [inline Sources/AWSLambdaRuntimeCore/Lambda.swift:92]: Should this be `NonInitializingByteBufferLambdaHandler`?

File: Sources/AWSLambdaRuntimeCore/LambdaRuntime.swift [hunk 3]
```diff
@@ -35,18 +37,41 @@ public final class LambdaRuntime<Handler: ByteBufferLambdaHandler> {
...
+            logger: logger,
+            configuration: .init()
+        )
     }
 
-    init(handlerType: Handler.Type, eventLoop: EventLoop, logger: Logger, configuration: LambdaConfiguration) {
+    /// Create a new `LambdaRuntime`.
+    ///
+    /// - parameters:
+    ///     - handlerProvider: A provider of the ``ByteBufferLambdaHandler``  the `LambdaRuntime` will manage.
```
Comment (tachyonics) [inline Sources/AWSLambdaRuntimeCore/LambdaRuntime.swift:60]: Same as the other comments.

**Comment by tachyonics** [comment]
Just some small comments about comments. Otherwise lgtm.

**Comment by tomerd** [comment]
@tachyonics new name idea: `LambdaRuntimeHandler`. wdyt?

File: Sources/AWSLambdaRuntimeCore/Lambda.swift [hunk 4]
```diff
@@ -85,6 +85,19 @@ public enum Lambda {
...
         configuration: LambdaConfiguration = .init(),
         handlerType: (some ByteBufferLambdaHandler).Type
+    ) -> Result<Int, Error> {
+        Self.run(configuration: configuration, handlerProvider: handlerType.makeHandler(context:))
+    }
+
+    /// Run a Lambda defined by implementing the ``LambdaRuntimeHandler`` protocol.
+    /// - parameters:
+    ///     - configuration: A Lambda runtime configuration object
+    ///     - handlerProvider: A provider of the ``LambdaRuntimeHandler``` to invoke.
```
Comment (tachyonics) [inline Sources/AWSLambdaRuntimeCore/Lambda.swift:95]: Extra tick...

**Comment by tomerd** [comment]
@fabianfett any concerns?

File: Sources/AWSLambdaRuntimeCore/Lambda.swift (hunk 4 above)
Comment (tomerd) [reply Sources/AWSLambdaRuntimeCore/Lambda.swift:95]:
```suggestion
    ///     - handlerProvider: A provider of the ``LambdaRuntimeHandler`` to invoke.
```

**Comment by fabianfett** [comment]
Tbh. I think this doesn't solve the actual issue. The lifetime handling in Lambda was designed and implemented before ServiceLifecycle came about. For this reason, using long running clients in Lambda currently feels absolutely weird.

For this reason, I think, we should reconsider what the Lambda API should look like taking into account that it should play nicely with ServiceLifecycle.

General ideas:

1. `LambdaRuntime` should implement `Service`
2. If a Lambda runs locally it should set up a `ServiceGroup` to handle signal handling
3. If a user needs long running Services within their Lambda they should be able to use ServiceGroups to manage their lifetimes.

**Pull request merged**